apps/aevm/src/aevm_eeevm_store.erl

```diff
@@ -196,31 +195,33 @@ store_maps(Maps0, Store) ->
 perform_update({new_inplace, NewId, OldId}, Store) ->
     OldKey   = <<OldId:256>>,
     NewKey   = <<NewId:256>>,
-    OldEntry = maps:get(OldKey, Store),
-    Store1   = maps:remove(OldKey, Store),
-    Store1#{ NewKey => OldEntry };
+    OldEntry = store_get(OldKey, Store),
+    %% Subtle: Don't remove the old entry because the mp trees requires there
+    %% to be a value at any node that we want to get a subtree for. We need
+    %% this for store_subtree.
+    %% Store1   = store_remove(OldKey, Store),
+    store_put(NewKey, OldEntry, Store);
 perform_update({insert, Id, Key, Val}, Store) ->
     RealId = real_id(Id, Store),
-    Store#{ <<RealId:256, Key/binary>> => Val };
+    store_put(<<RealId:256, Key/binary>>, Val, Store);
 perform_update({delete, Id, Key}, Store) ->
     RealId = real_id(Id, Store),
-    maps:remove(<<RealId:256, Key/binary>>, Store);
+    store_remove(<<RealId:256, Key/binary>>, Store);
 perform_update({new, Id, Map0}, Store) ->
     Map = aevm_eeevm_maps:flatten_map(Store, Id, Map0),
     RefCount = 0,   %% Set later
     Size     = Map0#pmap.size,
     Bin      = aeso_data:to_binary({Map#pmap.key_t, Map#pmap.val_t}),
-    Info = #{ <<Id:256>> => ?MapInfo(Id, RefCount, Size, Bin) },
-    Data = maps:from_list(
-            [ {<<Id:256, Key/binary>>, Val} || {Key, Val} <- maps:to_list(Map#pmap.data) ]),
-    maps:merge(Store, maps:merge(Info, Data));
+    Info = [{<<Id:256>>, ?MapInfo(Id, RefCount, Size, Bin)}],
+    Data = [ {<<Id:256, Key/binary>>, Val} || {Key, Val} <- maps:to_list(Map#pmap.data) ],
+    lists:foldl(fun({K, V}, S) -> store_put(K, V, S) end, Store, Info ++ Data);
 perform_update({gc, Id}, Store) ->
     RealId = real_id(Id, Store),
-    %% Remove map info entry
-    Store1 = maps:remove(<<Id:256>>, Store),
-    %% Remove all the data
-    maps:filter(fun(<<Id1:256, Key/binary>>, _) when Key /= <<>> -> Id1 /= RealId;
-                   (_, _) -> true end, Store1).
+    %% Remove map info entry. Also remove RealId which we kept around for mp
```

Comment:
Ah, ok. So this is supposed to fix the problem above. Is it guaranteed to be performed before setting the store on exit?

Reply:
No. See above.